server/sql_info_fetcher.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"archive/zip"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"runtime/pprof"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/session"
+	"github.com/pingcap/tidb/statistics/handle"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"github.com/pingcap/tidb/util/testkit"
+)
+
+type sqlInfoFetcher struct {
+	do *domain.Domain
+	s  session.Session
+}
+
+type tableNamePair struct {
+	DBName    string
+	TableName string
+}
+
+type tableNameExtractor struct {
+	names map[tableNamePair]struct{}
+}
+
+func (tne *tableNameExtractor) Enter(in ast.Node) (ast.Node, bool) {
+	if _, ok := in.(*ast.TableName); ok {
+		return in, true
+	}
+	return in, false
+}
+
+func (tne *tableNameExtractor) Leave(in ast.Node) (ast.Node, bool) {
+	if t, ok := in.(*ast.TableName); ok {
+		tp := tableNamePair{DBName: t.Schema.L, TableName: t.Name.L}
+		if _, ok := tne.names[tp]; !ok {
+			tne.names[tp] = struct{}{}
+		}
+	}
+	return in, true
+}
+
+func (sh *sqlInfoFetcher) zipInfoForSQL(w http.ResponseWriter, r *http.Request) {
+	reqCtx := r.Context()
+	sql := r.FormValue("sql")
+	pprofTimeString := r.FormValue("pprof_time")
+	timeoutString := r.FormValue("timeout")
+	var (
+		pprofTime int
+		timeout   int
+		err       error
+	)
+	if pprofTimeString != "" {
+		pprofTime, err = strconv.Atoi(pprofTimeString)
+	}
+	if err != nil {
+		serveError(w, http.StatusBadRequest, "invalid value for pprof_time")
+		return
+	}
+	if pprofTime < 5 {
+		serveError(w, http.StatusBadRequest, "pprof time is too short")
+	}
+	if timeoutString != "" {
+		timeout, err = strconv.Atoi(timeoutString)
+	}
+	if err != nil {
+		serveError(w, http.StatusBadRequest, "invalid value for timeout")
+		return
+	}
+	if timeout < pprofTime {
+		timeout = pprofTime
+	}
+	pairs, err := sh.extractTableNames(sql)
+	if err != nil {
+		serveError(w, http.StatusBadRequest, fmt.Sprintf("invalid SQL text, err: %v", err))
+		return
+	}
+	zw := zip.NewWriter(w)
+	defer func() {
+		terror.Log(zw.Close())
+	}()
+	for pair := range pairs {
+		jsonTbl, err := sh.getStatsForTable(pair)
+		if err != nil {
+			terror.Log(err)
+			continue
```

Comment:
Should we continue or just throw error out?

Reply:
can we return a message to indicate the failure of dumping stats for this table?

Reply:
write a file like `dbname.tblname.stats.err`?